### PR TITLE
feat(analytics): Analytics Feature

### PR DESCRIPTION
**Pull request by yquansah**
This PR addresses the feature of mainly providing a uniform way of querying analytics through Flipt's API. We have been getting feature requests such as this for some time now, and our focus here is to provide a uniform contract that can be implemented by many different OLAP stores which then sits behind an API that the frontend can consume.

## Main Features of this PR
1. Provide configuration definitions for analytics store
2. Provide analytics DDL for clickhouse in migrations directory
3. Provide an `analytics.proto` definition for giving an API on top of an analytics store
4. Integrate a clickhouse client for the first implementation of an analytics store
5. UI for showing analytics on Flag page (on new analytics tab)

### CODE REVIEW

**Comment by erka** [comment]
This looks great!

I have one question. How will `namespace_key` be passed around? Is it part of `flag_key`?

File: rpc/flipt/analytics/analytics.proto [hunk 1]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+
+package flipt.analytics;
+
+option go_package = "go.flipt.io/flipt/rpc/flipt/analytics";
+
+message GetFlagEvaluationsCountRequest {
+    string flag_key = 1;
```
Comment (markphelps) [inline rpc/flipt/analytics/analytics.proto:8]: I think we need `namespace_key` as @erka mentioned and in the REST path as well

File: internal/cmd/http.go [hunk 2]
```diff
@@ -146,6 +152,7 @@ func NewHTTPServer(
 		r.Mount("/api/v1", api)
 		r.Mount("/evaluate/v1", evaluateAPI)
 		r.Mount("/internal/v1", evaluateDataAPI)
+		r.Mount("/internal/analytics/v1", analyticsAPI)
```
Comment (markphelps) [inline internal/cmd/http.go:155]:
```suggestion
		r.Mount("/analytics/v1", analyticsAPI)
```

maybe?

or.. we could do:

```
/internal/v1/analytics
```

and just mount it above the `/internal/v1/` route so it matches first

File: internal/server/analytics/clickhouse/client.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	timeFormat            = "2006-01-02 15:04:05"
+)
+
+type Client struct {
+	conn *sql.DB
+}
+
+// New constructs a new clickhouse client that conforms to the analytics.Client contract.
+func New(logger *zap.Logger, connectionString string) (*Client, error) {
+	conn, err := connect(connectionString)
```
Comment (markphelps) [inline internal/server/analytics/clickhouse/client.go:32]: we might want to wrap the connection logic in a `sync.Once`

**Review by markphelps — COMMENTED** [review]
couple more minor things.. looking great tho!

File: internal/cmd/http.go (hunk 2 above)
Comment (yquansah) [reply internal/cmd/http.go:155]: @markphelps That is cool with me! We can do that

File: rpc/flipt/analytics/analytics.proto [hunk 4]
```diff
@@ -0,0 +1,20 @@
+syntax = "proto3";
+
+package flipt.analytics;
+
+option go_package = "go.flipt.io/flipt/rpc/flipt/analytics";
+
+message GetFlagEvaluationsCountRequest {
```
Comment (markphelps) [inline rpc/flipt/analytics/analytics.proto:7]: It may make sense to change this to `GetFlagEvaluationsStatsRequests/Response` to enable multiple statistics in the same RPC. I can think of `avg`, `p99`, etc.

Wdyt? Would it make sense to have a general stats endpoint that we can add new values to over time instead of having to add more methods for each stat? Or are counts different enough from histogram type metrics that we may need at least 2 different RPC calls?

Comment (GeorgeMac) [reply rpc/flipt/analytics/analytics.proto:7]: I would keep counts and duration quantiles as separate RPCs personally, as they're discretely separate measurements that will be served by separate queries.
This will be useful from an understanding the performance of these queries themselves in isolation from one another.
Otherwise, we could end up with one beast of an RPC call that does all the things.

File: internal/server/analytics/clickhouse/client.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+			return
+		}
+
+		connection, err := connect(connectionString)
+		if err != nil {
+			clickhouseErr = err
+			return
+		}
+
+		conn = connection
```
Comment (GeorgeMac) [inline internal/server/analytics/clickhouse/client.go:56]: We could probably establish the connection once, instead of reconnecting after performing migrations:
```suggestion
		conn, clickhouseErr = connect(connectionString)
		if clickhouseErr != nil {
			return
		}
		
		clickhouseErr = runMigrations(logger, conn)
		return
```

**Review by GeorgeMac — APPROVED** [review]
Change is looking great so far 💪

**Comment by GeorgeMac** [comment]
I wasn't paying attention and meant to make that a comment, not an approval review. But I do still approve of this PR 😂 so not wrong.

File: internal/server/evaluation/evaluation.go [hunk 6]
```diff
@@ -34,6 +34,13 @@ func (s *Server) Variant(ctx context.Context, r *rpcevaluation.EvaluationRequest
 		return nil, err
 	}
 
+	if s.analyticsStore != nil {
+		err := s.analyticsStore.IncrementFlagEvaluation(ctx, r.NamespaceKey, r.FlagKey)
+		if err != nil {
+			s.logger.Info("failure to add metric", zap.String("namespace_key", r.NamespaceKey), zap.String("flag_key", r.FlagKey), zap.Error(err))
+		}
+	}
+
```
Comment (yquansah) [inline internal/server/evaluation/evaluation.go:43]: I am going to do this action asynchronously as it is adding some considerable millisecond latency to the overall request.

**Comment by erka** [comment]
@yquansah I know I am a bit late with this request. Is it possible to add `reason` and `match` to analytics table in clickhouse? My thoughts are that it's great to show evaluation chart but if key is disabled that doesn't provide much value in to the end user and  for us. Also it could be great to know proportion of `match`/ `enabled` for the flag. If it's only true or false may provide better insight as may say it's a stale flag as there is no variation happen.

File: internal/server/analytics/clickhouse/client.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+	if err := conn.Ping(); err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (c *Client) GetFlagEvaluationsCount(ctx context.Context, req *analytics.GetFlagEvaluationsCountRequest) ([]string, []float32, error) {
+	fromTime, err := time.Parse(timeFormat, req.From)
```
Comment (GeorgeMac) [inline internal/server/analytics/clickhouse/client.go:97]: I think this is equivalent to the constant in the `time` package. You can drop the constant from this package too then.
```suggestion
	fromTime, err := time.Parse(time.DateTime, req.From)
```

File: internal/server/analytics/clickhouse/client.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	toTime, err := time.Parse(timeFormat, req.To)
```
Comment (GeorgeMac) [inline internal/server/analytics/clickhouse/client.go:102]:
```suggestion
	toTime, err := time.Parse(time.DateTime, req.To)
```

**Review by GeorgeMac — COMMENTED** [review]
I added some suggestions around the changes to the migrator package and just a couple clean up comments. Change is looking great though.

**Comment by yquansah** [comment]
@erka You are not late at all!! That is actually what is going to happen once I open it up. I plan to use https://clickhouse.com/docs/en/sql-reference/data-types/map to essentially have "tags" per measured metric.

File: internal/storage/sql/migrator.go [hunk 9]
```diff
@@ -34,13 +36,34 @@ type Migrator struct {
...
+	)
+
+	if analytics {
+		asql, adriver, err := openAnalytics(cfg, Options{migrate: true})
+		if err != nil {
+			return nil, fmt.Errorf("opening db: %w", err)
+		}
+
+		sql = asql
+		driver = adriver
```
[the remark this answers is not in the document]
Comment (yquansah) [reply internal/storage/sql/migrator.go]: @GeorgeMac That was the pattern I initially went with, then I ran into the the issue of how to DRY it up haha, but I think I figured it out! Thanks, I changed to using this pattern.

File: internal/server/analytics/clickhouse/client.go (hunk 5 above)
Comment (yquansah) [reply internal/server/analytics/clickhouse/client.go]: @GeorgeMac Got you. I was following the same pattern as the RDMS ones too but it does make sense to just use the connection once as you suggested.

File: internal/server/evaluation/server.go [hunk 10]
```diff
@@ -18,21 +18,44 @@ type Storer interface {
 	GetEvaluationRollouts(ctx context.Context, flag storage.ResourceRequest) ([]*storage.EvaluationRollout, error)
 }
 
+type AnalyticsStoreMutator interface {
+	IncrementBooleanFlagEvaluation(ctx context.Context, namespaceKey string, resp *evaluation.BooleanEvaluationResponse) error
+	IncrementVariantFlagEvaluation(ctx context.Context, namspaceKey string, resp *evaluation.VariantEvaluationResponse) error
```
Comment (markphelps) [inline internal/server/evaluation/server.go:23]:
```suggestion
	IncrementVariantFlagEvaluation(ctx context.Context, namespaceKey string, resp *evaluation.VariantEvaluationResponse) error
```

File: internal/server/analytics/clickhouse/client.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	return nil
+}
+
+func connect(connectionString string) (*sql.DB, error) {
+	conn := clickhouse.OpenDB(&clickhouse.Options{
+		Addr: []string{connectionString},
+		Auth: clickhouse.Auth{
+			Database: "flipt_analytics",
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/server/analytics/clickhouse/client.go:83]: I would assume the db name should be provided by the caller in the connection string

File: internal/config/analytics.go [hunk 12]
```diff
@@ -17,8 +18,18 @@ type AnalyticsConfig struct {
 
 // ClickhouseConfig defines the connection details for connecting Flipt to Clickhouse.
 type ClickhouseConfig struct {
-	Enabled bool   `json:"enabled,omitempty" mapstructure:"enabled" yaml:"enabled,omitempty"`
-	URL     string `json:"url,omitempty" mapstructure:"url" yaml:"url,omitempty"`
+	Enabled  bool   `json:"enabled,omitempty" mapstructure:"enabled" yaml:"enabled,omitempty"`
+	URL      string `json:"url,omitempty" mapstructure:"url" yaml:"url,omitempty"`
+	Username string `json:"username,omitempty" mapstructure:"username" yaml:"username,omitempty"`
```
Comment (markphelps) [inline internal/config/analytics.go:23]:
```suggestion
	Username string `json:"-" mapstructure:"username" yaml:"username,omitempty"`
```

File: internal/config/analytics.go [hunk 13]
```diff
@@ -17,8 +18,18 @@ type AnalyticsConfig struct {
...
+	Password string `json:"password,omitempty" mapstructure:"password" yaml:"password,omitempty"`
```
Comment (markphelps) [inline internal/config/analytics.go:24]:
```suggestion
	Password string `json:"-" mapstructure:"password" yaml:"password,omitempty"`
```

we don't want to expose these publicly over the `/meta/config` endpoint

File: ui/src/components/graphs/index.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+              },
+              title: {
+                display: true,
+                text: 'Time'
+              }
+            },
+            y: {
+              title: {
+                display: true,
+                text: 'Number of evaluations'
```
Comment (markphelps) [inline ui/src/components/graphs/index.tsx:72]:
```suggestion
                text: 'Evaluations'
```

File: ui/src/app/flags/Flag.tsx [hunk 15]
```diff
@@ -32,7 +33,8 @@ import {
 import Rollouts from './rollouts/Rollouts';
 const tabs = [
   { name: 'Variants', to: '' },
-  { name: 'Rules', to: 'rules' }
+  { name: 'Rules', to: 'rules' },
+  { name: 'Analytics', to: 'analytics' }
```
Comment (markphelps) [inline ui/src/app/flags/Flag.tsx:37]: the analytics tab seems to be missing for Boolean flag types or maybe with no rollout rules?

CleanShot 2024-02-05 at 09 37 33

File: ui/src/app/flags/analytics/Analytics.tsx [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+      values: getFlagEvaluationCount.data?.values
+    };
+  }, [getFlagEvaluationCount]);
+
+  const initialValues = {
+    durationValue: selectedDuration?.key
+  };
+
+  return (
+    <div className="mx-12 my-12">
```
Comment (markphelps) [inline ui/src/app/flags/analytics/Analytics.tsx:89]: I think we should show a 'Well' that says analytics are not enabled if they don't have any analytics backend configured, and then link to instructions on how to configure them (in docs see: FLI-825)

We do something similar for Auth Tokens if they don't have that method enabled: https://github.com/flipt-io/flipt/blob/2fbc00929d32ebf00cf6359a7c29a60d33ea83bf/ui/src/app/tokens/Tokens.tsx#L174-L193

File: internal/config/analytics.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/spf13/viper"
+)
+
+// AnalyticsConfig defines the configuration for various mechanisms for
+// reporting and querying analytical data for Flipt.
+type AnalyticsConfig struct {
+	Enabled    bool             `json:"enabled,omitempty" mapstructure:"enabled" yaml:"enabled,omitempty"`
```
Comment (markphelps) [inline internal/config/analytics.go:14]: do we need this? I asked previously, but if no analytics backends are configured I think we can not require this as an explicit config value?

File: internal/config/analytics.go [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+	Clickhouse ClickhouseConfig `json:"clickhouse,omitempty" mapstructure:"clickhouse" yaml:"clickhouse,omitempty"`
```
Comment (markphelps) [inline internal/config/analytics.go:15]: wdyt about nesting this under a subkey like `storage` or similar so that its not a 'top level' key

ie: `analytics.storage.clickhouse`

File: internal/config/analytics.go [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+	URL      string `json:"url,omitempty" mapstructure:"url" yaml:"url,omitempty"`
+	Username string `json:"username,omitempty" mapstructure:"username" yaml:"username,omitempty"`
+	Password string `json:"password,omitempty" mapstructure:"password" yaml:"password,omitempty"`
+}
+
+func (c *ClickhouseConfig) Auth() clickhouse.Auth {
+	return clickhouse.Auth{
+		Username: c.Username,
+		Password: c.Password,
+		Database: "flipt_analytics",
```
Comment (markphelps) [inline internal/config/analytics.go:31]: is the db part of the url? or do we 'own' the db name?

**Review by markphelps — CHANGES_REQUESTED** [review]

File: internal/config/analytics.go (hunk 17 above)
Comment (yquansah) [reply internal/config/analytics.go:14]: @markphelps Hmm yeah you are right. We can imply if it is enabled if any of the backends are enabled basically.

File: internal/config/analytics.go (hunk 19 above)
Comment (yquansah) [reply internal/config/analytics.go:31]: @markphelps Yeah the idea is that we 'own' the db name, basically that the users would have to create the DB beforehand.

I believe the DB could be part of the URL but the API that we are using separates these two concepts: https://pkg.go.dev/github.com/beebeeep/clickhouse-go/v2#Options

File: internal/config/analytics.go (hunk 18 above)
Comment (yquansah) [reply internal/config/analytics.go:14]: @markphelps We could. I guess that semantically makes more sense.

File: config/migrations/clickhouse/0_initial_clickhouse.up.sql [hunk 20]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE IF NOT EXISTS flipt_counter_analytics (
+    `timestamp` DateTime('UTC'), `analytic_name` String, `namespace_key` String, `flag_key` String, `flag_type` Enum('VARIANT_FLAG_TYPE' = 1, 'BOOLEAN_FLAG_TYPE' = 2), `reason` Enum('UNKNOWN_EVALUATION_REASON' = 1, 'FLAG_DISABLED_EVALUATION_REASON' = 2, 'MATCH_EVALUATION_REASON' = 3, 'DEFAULT_EVALUATION_REASON' = 4), `match` Bool, `value` UInt32
```
Comment (GeorgeMac) [inline config/migrations/clickhouse/0_initial_clickhouse.up.sql:2]: Your PR is already big enough, but I think we should make an issue to get an IT done, if there isn't an issue for that already.

One thing I think that could be easy to miss is extending these enums in the future.
If we fail to update this accordingly, we could cause failures to insert into CH.
Perhaps we can come up with a test case that will ensure that doesn't happen.

File: internal/server/analytics/clickhouse/mutation.go [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+		valueArgs = append(valueArgs, response.NamespaceKey)
+		valueArgs = append(valueArgs, response.FlagKey)
+		valueArgs = append(valueArgs, response.FlagType)
+		valueArgs = append(valueArgs, response.Reason)
+		valueArgs = append(valueArgs, response.Match)
+		valueArgs = append(valueArgs, 1)
+	}
+
+	//nolint:gosec
+	stmt := fmt.Sprintf("INSERT INTO %s VALUES %s", counterAnalyticsTable, strings.Join(valuePlaceHolders, ","))
```
Comment (GeorgeMac) [inline internal/server/analytics/clickhouse/mutation.go:34]: Just a passing thought, nothing to really do for now.
TL;DR something to keep an eye on r.e. performance.

I imagine constantly rebuilding this query to be quite wasteful, particularly under load. I remember you saying their native client is more performant for inserts, so maybe we just don't sweat this until we eventually move to their other client. However, one way to mitigate this might be cache the insert query based on buffer size. To try and amortize the cost and avoid constantly reallocating this query.

I wonder if CH has a better way to prepare this query, so that the placeholders don't have to be repeated in this way too. That might be something their own client takes advantage of.

Aside: I looked at this first, because of the injection problem, and the `nosec` there made suspicious too. But I see you're building the placeholder query with this. Which itself is just a series of palceholders, no user inputs, so it is safe. Then you're passing the values via the placeholder on exec, which is good. So I can see why you step around the linting. (I am assuming the noline rule is upset here because of fmt stringing a SQL ).

File: internal/server/analytics/sink_test.go [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+type sampleSink struct {
+	ch chan []*EvaluationResponse
+}
+
+func (s *sampleSink) IncrementFlagEvaluationCounts(ctx context.Context, responses []*EvaluationResponse) error {
+	go func() {
+		s.ch <- responses
+	}()
```
Comment (GeorgeMac) [inline internal/server/analytics/sink_test.go:24]: Curiosity: Does this need to be done in a goroutine, or does the otel trace provider already do this asynchronously from your code? My gut would've guessed that this is already performed in another goroutine and therefor won't block if you just send this inline. However, I could be mistaken.

If I am, you could also get away without doing this by giving the sample sink channel a non-zero buffer. That is a take it or leave it though.

File: internal/server/middleware/grpc/middleware.go [hunk 23]
```diff
@@ -88,33 +90,127 @@ type ResponseDurationRecordable interface {
...
+						}
+					}
+				case *evaluation.BooleanEvaluationResponse:
+					if evaluationRequest, ok := req.(*evaluation.EvaluationRequest); ok {
+						evaluationResponses := []*analytics.EvaluationResponse{
+							{
+								NamespaceKey: evaluationRequest.GetNamespaceKey(),
+								FlagKey:      r.GetFlagKey(),
+								FlagType:     evaluation.EvaluationFlagType_BOOLEAN_FLAG_TYPE.String(),
+								Match:        r.GetEnabled(),
```
Comment (GeorgeMac) [inline internal/server/middleware/grpc/middleware.go:152]: Im wondering if we should **not** set this field for boolean type flag.
Instead, I think we should add a `value` to evaluation response.

Variant can provide the `match` property, but it will set `value` to the matched variant.
Boolean will leave `match` unset, but set `value` to `enabled`.

File: internal/server/middleware/grpc/middleware.go [hunk 24]
```diff
@@ -88,33 +90,127 @@ type ResponseDurationRecordable interface {
...
+								Reason:       r.GetReason().String(),
+								Timestamp:    r.GetTimestamp().AsTime(),
+							},
+						}
+
+						if evaluationResponsesBytes, err := json.Marshal(evaluationResponses); err == nil {
+							keyValue := []attribute.KeyValue{
+								{
+									Key:   "flipt.evaluation.response",
+									Value: attribute.StringValue(string(evaluationResponsesBytes)),
```
Comment (GeorgeMac) [inline internal/server/middleware/grpc/middleware.go:139]: Here is a bit of a mad hack: https://pkg.go.dev/go.opentelemetry.io/otel/attribute#Stringer

You could add a `String() string` method to `*analytics.EvaluationResponse` that marshals out a Go string. Then instead of marshalling this to JSON and then unmarshalling again in the exporter, you could pass it here with `attributes.Stringer` and just cast it back to a `*analytics.EvaluationResponse` in the exporter.
Otel will take care of marshalling lazilly when it needs to go over the wire.

File: internal/config/analytics.go (hunk 19 above)
Comment (yquansah) [reply internal/config/analytics.go:39]: So it looks like the provide a [ParseDSN](https://pkg.go.dev/github.com/beebeeep/clickhouse-go/v2#ParseDSN) function that will take care of providing the clickhouse options for you. I might go that route instead.

But within the `Options` itself, they separate the concept of `Host` and `Username`/`Password`.

If you provide a username and password for an `Addr`, it actually errors out.

File: internal/server/middleware/grpc/middleware.go (hunk 24 above)
Comment (yquansah) [reply internal/server/middleware/grpc/middleware.go:139]: @GeorgeMac Ahh I started to implement it this way just now and realized,

What is actually happening here is I am marshalling a `[]*EvaluationResponse` to JSON instead of just a single `*EvaluationResponse`.

I chose to do it this way because it helps with the `Batch` implementation.

File: internal/server/analytics/clickhouse/mutation.go (hunk 21 above)
Comment (yquansah) [reply internal/server/analytics/clickhouse/mutation.go:34]: @GeorgeMac Yeah exactly! I tried to see if there was a better way to prepare this query. But you are right, we can look into that more native client.

I can place an issue in the project to do that. We can see if it has more features that is better suited for inserts.

File: internal/server/middleware/grpc/middleware.go (hunk 23 above)
Comment (yquansah) [reply internal/server/middleware/grpc/middleware.go:152]: @GeorgeMac Ahh this actually makes more sense! I forgot that Boolean's enabled property is the `value` itself

File: internal/server/analytics/sink_test.go (hunk 22 above)
Comment (yquansah) [reply internal/server/analytics/sink_test.go:24]: @GeorgeMac Yeah so OTEL actually calls `ExportSpans` synchronously: https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#SpanExporter

Which in turn would call this function. So when `span.End()` is called here: https://github.com/flipt-io/flipt/blob/337b179050f7fe2f14a75839aa79b5e49d4ed016/internal/server/analytics/sink_test.go#L63

If there is no channel waiting to receive at the time it is invoked it will block.

File: config/migrations/clickhouse/0_initial_clickhouse.up.sql (hunk 20 above)
Comment (yquansah) [reply config/migrations/clickhouse/0_initial_clickhouse.up.sql]: @GeorgeMac Yup will create an issue for this.

File: internal/server/middleware/grpc/middleware.go (hunk 24 above)
Comment (GeorgeMac) [reply internal/server/middleware/grpc/middleware.go:145]: Could you re-type that slice for the same ends?
```go
type EvaluationResponses []*EvaluationResponse

func (r EvaliationResponses) String() string {
    v, _ := json.Marshal(r)
    return string(v)
}
```

And then use `analytics.EvaluationResponses{}` instead of `[]*analytics.EvaluationResponse`?
Either way, not important, just seeing if we can avoid marshalling back and forth.

File: internal/config/analytics.go [hunk 25]
```diff
@@ -22,22 +24,25 @@ type AnalyticsStorageConfig struct {
...
-		Username: c.Username,
-		Password: c.Password,
-		Database: "flipt_analytics",
+// Options returns the connection option details for Clickhouse.
+func (c *ClickhouseConfig) Options() (*clickhouse.Options, error) {
+	options, err := clickhouse.ParseDSN(c.URL)
+	if err != nil {
+		return nil, err
+	}
+	if options.Auth.Database != "flipt_analytics" {
```
Comment (markphelps) [inline internal/config/analytics.go:41]: i think we want to allow them to use whatever db name they choose

File: cmd/flipt/migrate.go [hunk 26]
```diff
@@ -21,21 +57,23 @@ func newMigrateCommand() *cobra.Command {
...
+					return err
+				}
 			}
 
 			return nil
 		},
 	}
 
 	cmd.Flags().StringVar(&providedConfigFile, "config", "", "path to config file")
+	cmd.Flags().StringVar(&database, "database", "regular", "string to denote which database type to migrate")
```
Comment (markphelps) [inline cmd/flipt/migrate.go:77]:
```suggestion
	cmd.Flags().StringVar(&database, "database", "default", "string to denote which database type to migrate")
```

File: internal/config/analytics.go [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+//nolint:unparam
+func (a *AnalyticsConfig) setDefaults(v *viper.Viper) error {
+	v.SetDefault("analytics", map[string]any{
+		"enabled": "false",
+		"clickhouse": map[string]any{
+			"enabled": "false",
+			"url":     "",
+		},
```
Comment (markphelps) [inline internal/config/analytics.go:55]:
```suggestion
	v.SetDefault("analytics", map[string]any{
		"storage.clickhouse": map[string]any{
			"enabled": "false",
			"url":     "",
		},
```

File: internal/config/analytics.go (hunk 25 above)
Comment (yquansah) [reply internal/config/analytics.go:41]: @markphelps We could do that. They would have to create that DB beforehand. I guess we can clearly state that in our docs

Comment (markphelps) [reply internal/config/analytics.go:41]: can we try and parse the db name from the connection string?

if its there, then we use that, if not then we use this default?

Comment (GeorgeMac) [reply internal/config/analytics.go]: Yeah I think we definitely want them to create it first themselves. Like we do with the relational DB.

File: internal/server/analytics/clickhouse/mutation.go [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+// IncrementFlagEvaluationCounts does a batch insert of flag values from evaluations.
+func (c *Client) IncrementFlagEvaluationCounts(ctx context.Context, responses []*analytics.EvaluationResponse) error {
+	if len(responses) < 1 {
+		return nil
+	}
+
+	valuePlaceHolders := make([]string, 0, len(responses))
+	valueArgs := make([]interface{}, 0, len(responses)*9)
```
Comment (markphelps) [inline internal/server/analytics/clickhouse/mutation.go:19]:
```suggestion
	valueArgs := make([]interface{}, 0, len(responses)*COLUMNS)
```

Could we make `9` a constant so its not a magic number

File: ui/src/app/flags/analytics/Analytics.tsx [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+    flagKey: flag.key,
+    from: format(
+      addMinutes(
+        addMinutes(
+          nowISO,
+          selectedDuration?.value ? selectedDuration.value * -1 : -60
+        ),
+        nowISO.getTimezoneOffset()
+      ),
+      timeFormat
```
Comment (markphelps) [inline ui/src/app/flags/analytics/Analytics.tsx:74]: would it be possible to show the evaluation timestamps in the format that the user selects in preferences? ie UTC vs local?

CleanShot 2024-02-07 at 10 12 30

File: ui/src/app/flags/analytics/Analytics.tsx [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+    ),
+    to: format(addMinutes(nowISO, nowISO.getTimezoneOffset()), timeFormat)
+  });
+
+  const flagEvaluationCount = useMemo(() => {
+    const fetchError = getFlagEvaluationCount.error as FetchBaseQueryError;
+    return {
+      timestamps: getFlagEvaluationCount.data?.timestamps,
+      values: getFlagEvaluationCount.data?.values,
+      unavailable: fetchError?.status === 501
```
Comment (markphelps) [inline ui/src/app/flags/analytics/Analytics.tsx:84]: another option is we could check the `/meta/config` endpoint to see if analytics are enabled before making the call and depending on a 501

CleanShot 2024-02-07 at 10 18 46

I Could do this in a future PR though as it will require some redux work I think

**Review by markphelps — COMMENTED** [review]
couple minor last minute comments, otherwise lgtm!

**Comment by markphelps** [comment]
One other thing I noticed is that the `schema_migrations` table seems to be marking the migration as `dirty` after first migrate? I wonder if this is due to us potentially running migrate multiple times or something related to how golang migrate works?

I'm gonna check the sqlite db to see if the same thing happens there

CleanShot 2024-02-07 at 10 31 42

File: ui/src/components/graphs/index.tsx [hunk 31]
```diff
@@ -33,11 +38,22 @@ type BarGraphProps = {
...
 
 export function BarGraph({ flagKey, timestamps, values }: BarGraphProps) {
+  const timezone = useSelector(selectTimezone);
+  const { inTimezone } = useTimezone();
+  const formattedTimestamps = timestamps.map((timestamp) =>
+    inTimezone(timestamp).slice(0, -4)
+  );
+  const isUTC = useMemo(() => timezone === Timezone.UTC, [timezone]);
+
+  console.log(formattedTimestamps);
```
Comment (markphelps) [inline ui/src/components/graphs/index.tsx:48]:
```suggestion
```

**Review by markphelps — COMMENTED** [review]
one minor console.log removal, then ship it!

**Review by markphelps — APPROVED** [review]
📈
